contracts/ScheduledPaymentModule.sol

```diff
@@ -78,6 +97,54 @@ contract ScheduledPaymentModule is Module {
         emit ScheduledPaymentCancelled(spHash);
     }
 
+    function executeScheduledPayment(
+        address token,
+        uint256 amount,
+        address payee,
+        uint256 maxGasPrice,
+        address gasToken,
+        uint256 _nonce,
+        uint256 payAt,
+        uint256 gasPrice
+    ) external onlyCrank {
+        bytes32 spHash = createSpHash(
+            token,
+            amount,
+            payee,
+            maxGasPrice,
+            gasToken,
+            _nonce,
+            payAt
+        );
+        if (!spHashes.contains(spHash)) revert UnknownHash(spHash);
+        if (block.timestamp < payAt.add(1 minutes))
+            revert InvalidPeriod(spHash);
+        if (gasPrice > maxGasPrice) revert ExceedMaxGasPrice(spHash);
+
+        // execTransactionFromModule to execute the sheduled payment
+        bytes memory spTxData = abi.encodeWithSelector(
+            0xa9059cbb,
+            payee,
+            amount
+        );
+        uint256 gasUsed = gasleft();
+        if (!exec(token, 0, spTxData, Enum.Operation.Call))
+            revert PaymentExecutionFailed(spHash);
+        gasUsed = gasUsed.sub(gasleft());
+
+        // execTransactionFromModule for gas reimbursement
+        bytes memory gasTxData = abi.encodeWithSelector(
+            0xa9059cbb,
+            IConfig(config).getFeeReceiver(),
+            gasUsed.mul(gasPrice)
```

Comment:
what about the gas necessary for all the calls that happen after `gasUsed` is determined? gnosis safe contracts use an idea called `baseGas` to account for this additional gas cost, which is fixed. perhaps we should also have a `baseGas` consideration for all the calls that happen after `gasUsed` is calculated. Keep in mind that there is also the gas cost necessary transfer tokens out of the safe to even pay for the gas.

Reply:
note that the gas used for transferring the reimbursement of gas costs is a bit of a chicken and the egg--this is why the gnosis safe relay server does all the gas estimation using intentional reverts--its a very ingenious solution to the problem. measuring with actual on-chain calls, but reverting at the last moment with the gas used as the payload of the revert message so that the measurement is is non-state altering (and hence free to perform)

Reply:
here is an example of how gnosis does this--this function is called by the relay server to estimate gas, and the relay server parses the error message to understand host much gas was consumed. we could do something like this to wrap everything in this method, and that could be passed as an additional param to the spHash--the gasUsed https://github.com/safe-global/safe-contracts/blob/main/contracts/GnosisSafe.sol#L306-L327

Reply:
@habdelra, I added the estimate function (`estimateExecutionGas`). This function will only estimate the safe transaction execution part, not the whole part of the `executeScheduledPayment` function. That's why besides executionGas, I also add baseGas to the hash.

Reply:
That makes sense—unlike the real gnosis relay server you won’t know the gas price to use for a future txn, which means you won’t be able to estimate a token xfer for gas reimbursement since you don’t know the amount to xfer. 

I wonder though, if the gas limit is generally constant for any token xfer amount? If so then maybe you can just use a hard coded amount for estimation purposes, the key being that you are actually using the gas token’s contract with the assumption that the ledger change necessary for a token transfer is constant regardless of the token transfer amount. 

Otherwise, the baseGas will have to be token specific. And also I guess the baseGas also probably has the very same assumption, that it’s the same gas limit regardless of the token amount being transferred.

Reply:
As far as I know, the gas limit is generally constant for any token transfer amount, so I agree with you we can put the baseGas hardcoded in the code.